dlock/dlock.go

```diff
@@ -0,0 +1,150 @@
+// Package dlock is a distributed lock to enable advanced synchronization for Mattermost Plugins.
+//
+// if you're new to distributed locks and Mattermost Plugins please read sample use case scenarios
+// at: https://community.mattermost.com/core/pl/bb376sjsdbym8kj7nz7zrcos7r
+package dlock
+
+import (
+	"context"
+
+	"sync"
+	"time"
+
+	pluginapi "github.com/lieut-data/mattermost-plugin-api"
+	"github.com/pkg/errors"
+)
+
+const (
+	// storePrefix used to prefix lock related keys in KV store.
+	storePrefix = "dlock:"
+)
+
+const (
+	// lockTTL is lock's expiry time.
+	lockTTL = time.Second * 15
+
+	// lockRefreshInterval used to determine how long to wait before refreshing
+	// a lock's expiry time.
+	lockRefreshInterval = time.Second
+
+	// lockTryInterval used to wait before trying to obtain the lock again.
+	lockTryInterval = time.Second
+)
+
+// API is a portion of pluginapi to keep locks' state and log errors.
+type API interface {
```

Comment:
I think it's best to split this interface into two, as Logger and Store and accept them separately at New(). That way it's easier for caller to pass necessary implementations and to avoid possible name conflicts inside possible big variations of API interfaces. I suggest this as an idiomatic pattern for plugins.

This can be accomplished in a separate PR.